core/shared/src/main/scala/org/typelevel/log4cats/Log.scala

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2018 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.log4cats
+
+import scala.collection.mutable
+import scala.concurrent.duration.FiniteDuration
+
+/**
+ * Low-level interface exposing methods to enrich a log record with relevant information. The
+ * methods are designed to capture elements that cannot be easily captured from a monadic context
+ * (or by running an effect). Elements such as timestamps should be provided by means of
+ * middlewares.
+ */
+trait Log[Ctx] {
+  def timestamp: Option[FiniteDuration]
+  def level: KernelLogLevel
+  def message: String
```

Comment:
This should probably be lazy all the way to the edge.

```suggestion
  def message: () => String
```